### PR TITLE
Diagonal Difference for Square Matrices

### PR DESCRIPTION
**Pull request by jremes-foss**
#### Brief description of what is fixed or changed

Introduces a diagonal difference for square matrix (absolute difference between the sums of its diagonals).

#### Release Notes

- matrices
    - Introduces diagonal_difference method to sympy.matrices.common

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
I've never heard of the diagonal difference. Is it useful for something?

**Comment by Newman101** [comment]
@oscarbenjamin it's occasionally useful for some numerical analysis. Not the most common thing, but it does exist.

File: sympy/matrices/tests/test_commonmatrix.py [hunk 1]
```diff
@@ -1257,6 +1257,10 @@ def test_diagonal():
     assert banded({i: list(M.diagonal(i))
         for i in range(1-M.rows, M.cols)}) == M
 
+def test_diagonal_difference():
+    m = Matrix(3, 3, range(9))
+    diag_diff = m.diagonal_difference()
+    assert diag_diff == 2
```
Comment (sylee957) [inline sympy/matrices/tests/test_commonmatrix.py]:
```suggestion
    assert diag_diff == 0
```

**Comment by oscarbenjamin** [comment]
Does this need to be a method or could it be a utility function somewhere? Do we even want this at all?

We already have a lot of matrix classes that have a lot of methods and this seems obscure. I'm not sure that we should extend the API for this when it isn't common and can be done as a one-liner:
```
diagonal_difference = Abs(trace(a) - trace(a[::-1,:]))
```
Googling for "diagonal difference" just shows that this is a popular programming exercise - I haven't found any use for it.

**Comment by sylee957** [comment]
I also suppose that the term `Diagonal Difference` is most likely to be invented from some coding challenges, but in my opinion, I'm open to merging this even if it looks too trivial for the first place.

I think this can be made more sophisticated later, if we can have some separate knowledgebase for the computation of `diagonal_difference` for special matrices, like how we provide shortcuts for computing the `trace` for some special matrices.

I think that SymPy still lacks much knowledgebase for the magic square problems.

**Comment by oscarbenjamin** [comment]
Is there a better name than diagonal difference? I can't find any mathematical information using that term.

If we want this to be more efficient e.g. for sparse matrices then can we not just use `trace` and slicing like I showed above? Each of those can be made efficient for sparse matrices without needing to make multiple implementations for diagonal difference.

**Comment by Newman101** [comment]
Hi. I'm fine both ways - if this does not get merged, then it would be a good idea at least to add sum of anti-diagonal. I can make a separate pull request for this or refactor this PR. It's up to you.

**Comment by oscarbenjamin** [comment]
Isn't the sum of antidiagonal just `trace(a[::-1,:])`?

I don't necessarily object to this being added but I would like to see some justification. You said above that "it does exist" but I don't see any reference to it anywhere. Does numpy or anything else have a method/function for this? If so what is it called?

**Comment by Newman101** [comment]
There is the diagonal difference table in numerical analysis. After doing some research, I came to conclusion that this is way too obscure. Hence I close this pull request.

**Comment by oscarbenjamin** [comment]
@Newman101 I think you're probably right that this is too obscure. However thank you for taking the time to open a pull request and I hope that this doesn't discourage you from opening other pull requests in future. If you are considering a feature you can also open an issue to discuss the idea first.

**Comment by Newman101** [comment]
@oscarbenjamin Hi. No problem at all - this does not discourage me from opening other pull requests in the future. If there is a feature which pops into my mind, I'll do my research beforehand.

**Pull request closed without merging**